### DISCUSSION TITLE
Best practice of error handling?

### DISCUSSION THREAD

**Discussion by Reverier-Xu** [rune-rs/rune#716] [Q&A] [answered]
Hi, I'm trying to embed rune in my project as script engine, I found that it is difficult to handle errors.

I have several modules that contain functions returning different type of Result, such as

```rust
#[rune::function]
pub fn func_a() -> Result<(), AError> {...}

#[rune::function]
pub fn func_b() -> Result<(), BError> {...}
```

and use them in the script:

```rust
pub fn main() {
    func_a()?;
    func_b()?;
    return sth;
}
```

```rust
    let output = vm.async_call(["main"], (args)).await?;
    let object: Object = rune::from_value(output)?;
```

I found that if `func_a` or `func_b` triggered error and return early from script, the final output will raise an type convert error ``Expected data of type `Object`, but found `XXXError` ``, I tried turn the type from `Object` to `Result<Object, AError>` then `func_a` works but `func_b` not.

I also tried change the convert type from `Object` to `anyhow::Result<Object>`, but it also raised an error `` Expected data of type `anyhow::Error`, but found `Error` ``. although it is enough to detect whether an error is encountered in script, I still want to get the error that returns from script instead of type conversion errors.

What is the best practices of error handling when integrating rune? thanks for any response!

**Comment by udoprog** [comment, marked as answer]
You have two options on the edge, either everything needs to be a `Result` so you can convert it into a `std::result::Result<T, E>`, or you'll have to typecheck by comparing the value of [`Value::type_hash`](https://docs.rs/rune/latest/src/rune/runtime/value.rs.html#1122-1159) with the corresponding type hash and only perform the conversion if they match.

The former is what's encouraged. Try to keep your types coherent, at least at the edge of your function. In your case that would mean returning `Ok(sth)` instead of just `sth`.

The latter is *possible*, but won't really be easy to work with since calculating a type hash is a bit tricky right now. In git we actually have a macro that can do it which will be in the next release.
